engine/widget/region.3
.\"	$Csoft: region.3,v 1.9 2003/03/05 18:22:18 vedge Exp $
.\"
.\" Copyright (c) 2002, 2003 CubeSoft Communications, Inc.
.\" <http://www.csoft.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd August 21, 2002
.Dt REGION 3
.Os
.ds vT Agar API Reference
.ds oS Agar 1.0
.Sh NAME
.Nm region
.Nd agar widget container
.Sh SYNOPSIS
.Bd -literal
#include <engine/widget/window.h>
.Ed
.Sh DESCRIPTION
The
.Nm
structure is a simple widget container.
It is responsible for positioning and sizing a list of child widgets.
.Sh INITIALIZATION
.nr nS 1
.Ft struct region *
.Fn region_new "struct window *parent" "int flags" "int rx" \
               "int ry" "int rw" "int rh"
.Pp
.Ft void
.Fn region_init "struct region *region" "int flags" "int rx" \
                "int ry" "int rw" "int rh"
.Pp
.Ft void
.Fn region_attach "struct region *parent" "struct widget *child"
.Pp
.Ft void
.Fn region_detach "struct region *parent" "struct widget *child"
.Pp
.nr nS 0
.Pp
The
.Fn region_new
function allocates, initializes, and attaches a
.Nm
structure.
The
.Fn region_init
function initializes a
.Nm
structure.
The
.Fa rx
and
.Fa ry
arguments define the position of the region inside the window, and may have the
following meanings:
.Pp
.Bl -tag -compact -width ">= 0"
.It >= 0
A value in % scaled to the window size.
.It -1
With
.Dv REGION_HALIGN ,
position at the left of the last region.
With
.Dv REGION_VALIGN ,
position at the bottom of the last region.
.El
.Pp
The
.Fa rw
and
.Fa rh
arguments define the size of the region, and may have the following meanings:
.Pp
.Bl -tag -compact -width ">= 0"
.It > 0
A value in % scaled to the view size.
.It 0
With
.Dv REGION_HALIGN ,
all remaining space at the right of the last region.
With
.Dv REGION_VALIGN ,
all remaining space at the bottom the last region.
.El
.Pp
The
.Fa flags
may include:
.Pp
.Bd -literal
#define REGION_HALIGN	0x01	/* Align widgets horizontally */
#define REGION_VALIGN	0x02	/* Align widgets vertically */
#define REGION_CLIPPING	0x04	/* Set the clipping rectangle to the region
			 	   area before drawing */
.Ed
.Pp
The
.Fn region_attach
function attaches a widget to the region.
.Fn region_detach
detaches a widget from the region, without freeing it.
.Sh EVENTS
The
.Nm
object sends its widgets the following events:
.Pp
.Bl -tag -width 2n
.It Fn attached "struct region *parent"
Widget attached to region
.Fa parent .
.It Fn detached "struct region *parent"
Widget detached from region
.Fa parent .
.El
.Sh SEE ALSO
.Xr agar 3 ,
.Xr view 3 ,
.Xr widget 3 ,
.Xr window 3
.Sh HISTORY
The
.Nm
widget first appeared in Agar 1.0.


CHANGELOG.md
# Changelog
All notable changes to Agar will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.8.0] - Unreleased
### Added
- [**Style Editor**](https://libagar.org/man3/AG_StyleEditor): Display computed style rules in addition to per-instance overrides.
- [**AG_Button**](https://libagar.org/man3/AG_Button): New style attributes `text-align` and `vertical-align`.
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Introduce the `style()` method, called by the style compiler to parse widget-specific style attributes into an efficient binary representation.
- [**AG_Object**](https://libagar.org/man3/AG_Object): New function `AG_AccessVariable_NoLock()` (lockless variant of `AG_AccessVariable()`).
- [**AG_Object**](https://libagar.org/man3/AG_Object): Document the `AG_OBJECT_BOUND_EVENTS` flag.

### Removed
- [**AG_Window**](https://libagar.org/man3/AG_Window): Remove the "window-detached" event. It is redundant since the generic "detached" event does the same job.

### Changed
- Major performance improvements in the style compiler. Since it runs whenever widgets are resized, this should lead to better fluidity and overall lower CPU usage when Agar windows and container widgets are resized.
- [**AG_Object**](https://libagar.org/man3/AG_Object): In `AG_ObjectDestroy()`, do not call `AG_ObjectDetach()` on child objects. Instead, just cancel any active timers and null their parent pointers without raising a "detached" event. This allows us to perform garbage collection more efficiently.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): Make `AG_TLIST_NO_BGLINES` the default.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): In `AG_TlistNewPolledMs()`, accept a `ms` argument of 0 to disable periodical updates (the caller may still trigger updates with "tlist-poll").
- In `AG_WidgetCopyStyle()`, prevent redundant locking by using `AG_AccessVariable_NoLock()`.

### Fixed
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Fix redundant "detached" handler invocations. Improve overall efficiency of the garbage-collection step after windows are destroyed.
- [**AG_Console**](https://libagar.org/man3/AG_Console): Prevent a redundant lock operation in the backend of `AG_ConsoleMsg()`.

## [1.7.1] - 2025-11-15
### Added
- An awesome new Agar logo. Thanks [eightbit](https://8bitprodigy.github.io)!
- Added support for the [Cmake](https://cmake.org) build system.
- Added support for the [Meson](https://mesonbuild.com) build system. Thanks [coleman](https://git.sr.ht/~coleman/)!
- New functions [**AG_InitVideoSDL2**](https://libagar.org/man3/AG_InitVideoSDL2) and [**AG_SetVideoSurfaceSDL2**](https://libagar.org/man3/AG_SetVideoSurfaceSDL2) for integrating with an existing SDL2 display context. Thanks Brigham Keys!
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Add "Any File" type.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New function `AG_TlistCopy()`. Copy all items from a source to a destination `AG_Tlist`.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): New constructor `AG_ComboNewFn()`. Sets a `combo-expanded` handler.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): New member `nVisItems`. Set the number of items to show by default in expansions.
- [**AG_Checkbox**](https://libagar.org/man3/AG_Checkbox): New functions `AG_CheckboxText()` and `AG_CheckboxTextS()` to update the text label.
- Tweak the default color scheme. Document style defaults as a comment block in `gui/style.css`.
- New Ada bindings for the `Window`, `Box`, `Button`, `Checkbox` and `Combo` widget classes. Updated `ada/gui/demo`.

### Removed
- [**AG_Db**](https://libagar.org/man3/AG_Db): Remove support for DB5.
- [**AG_Window**](https://libagar.org/man3/AG_Window): Remove an unnecessary counter variable in `SizeAllocate()`.

### Changed
- Small cosmetic changes to the default color scheme. Set default font size back to 14pt.
- [**AG_Titlebar**](https://libagar.org/man3/AG_Titlebar): Enable the `AG_BUTTON_EXCL` optimization on titlebar buttons.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): Allow the dropdown list to be populated by a `combo-expanded` handler.
- [**AG_UCombo**](https://libagar.org/man3/AG_UCombo): Allow the dropdown list to be populated by a `ucombo-expanded` handler.

### Fixed
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): Make it again possible to statically initialize `list` before `combo-expanded`. Restores compatibility pre-1.6. Thanks Wally!
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Fix widget geometries not updating when switching to a different Type filter.
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Fix compilation where `GLOB_TILDE` is unavailable.
- SDL2 drivers: Require at least version 2.0.22 of SDL2 (`SDL_HINT_MOUSE_AUTO_CAPTURE` is necessary for best user experience).
- [**AG_ProgressBar**](https://libagar.org/man3/AG_ProgressBar): Make `padding` work as expected in progress bar. Thanks scaramacai!
- [**cocoa**](https://libagar.org/man3/AG_DriverCocoa): Fix a compilation error.
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Fix a keyboard handling error. Don't pass `KeyRelease` events to `XwcLookupString()`. Fixes issue #74. Thanks @tsesani and @scaramacai!
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Introduce the `no-im` (disable input methods) driver option.
- [**MAP**](https://libagar.org/man3/MAP): Fix a crash when attempting to create an object if `pLibsRoot` is NULL.
- [**AG_Web**](https://libagar.org/man3/AG_Web): Fix inline compilation problem due to a typo in `net/web.h`.
- On Darwin platforms, build with `_DARWIN_C_SOURCE` globally. Fixes issue #77. Thanks @kwhr0!
- On FreeBSD check for availability of `CLOCK_SECOND` in addition to `HAVE_CLOCK_GETTIME`.
- Test for the `dirfd()` function at compilation.

engine/widget/region.c
/*	$Csoft: region.c,v 1.22 2002/11/17 23:13:59 vedge Exp $	*/

/*
 * Copyright (c) 2002 CubeSoft Communications, Inc. <http://www.csoft.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistribution of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Neither the name of CubeSoft Communications, nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <engine/engine.h>

#include <engine/view.h>

#include <math.h>

#include "text.h"
#include "widget.h"
#include "window.h"

static const struct object_ops region_ops = {
	region_destroy,
	NULL,		/* load */
	NULL		/* save */
};

struct region *
region_new(void *parent, int flags, int rx, int ry, int rw, int rh)
{
	struct window *win = parent;
	struct region *reg;

	OBJECT_ASSERT(parent, "window");

	reg = emalloc(sizeof(struct region));
	region_init(reg, flags, rx, ry, rw, rh);
	window_attach(win, reg);
	return (reg);
}

void
region_init(struct region *reg, int flags, int rx, int ry, int rw, int rh)
{
	static pthread_mutex_t curreg_lock = PTHREAD_MUTEX_INITIALIZER;
	static int curreg = 0;
	char *name;

	pthread_mutex_lock(&curreg_lock);
	name = object_name("window-region", curreg++);
	pthread_mutex_unlock(&curreg_lock);

	object_init(&reg->obj, "window-region", name, NULL, 0, &region_ops);
	free(name);

	reg->flags = (flags != 0) ? flags : REGION_HALIGN;
	reg->rx = rx;
	reg->ry = ry;
	reg->rw = rw;
	reg->rh = rh;
	reg->x = 0;
	reg->y = 0;
	reg->w = 0;
	reg->h = 0;
	reg->spacing = 5;
	reg->win = NULL;
	TAILQ_INIT(&reg->widgetsh);
}

void
region_destroy(void *p)
{
	struct region *reg = p;
	struct widget *wid, *nextwid;

	for (wid = TAILQ_FIRST(&reg->widgetsh);
	     wid != TAILQ_END(&reg->widgetsh);
	     wid = nextwid) {
		nextwid = TAILQ_NEXT(wid, widgets);
		event_post(wid, "detached", p);		/* Notify */
		object_destroy(wid);			/* Free */
	}
	TAILQ_INIT(&reg->widgetsh);
}

/* Attach a widget to this region. */
void
region_attach(void *parent, void *child)
{
	struct region *reg = parent;
	struct widget *wid = child;

	OBJECT_ASSERT(parent, "window-region");
	OBJECT_ASSERT(child, "widget");
	
	wid->win = reg->win;
	wid->reg = reg;

	pthread_mutex_lock(&reg->win->lock);
	TAILQ_INSERT_TAIL(&reg->widgetsh, wid, widgets);	/* Attach */
	pthread_mutex_unlock(&reg->win->lock);
	
	OBJECT(child)->state = OBJECT_CONSISTENT;
	event_post(child, "attached", "%p", parent);		/* Notify */
}

/* Detach a widget from this region. */
void
region_detach(void *parent, void *child)
{
	struct region *reg = parent;
	struct widget *wid = child;

	OBJECT_ASSERT(parent, "window-region");
	OBJECT_ASSERT(child, "widget");

	pthread_mutex_lock(&reg->win->lock);
	TAILQ_REMOVE(&reg->widgetsh, wid, widgets);		/* Detach */
	pthread_mutex_unlock(&reg->win->lock);
	
	event_post(child, "detached", parent);			/* Notify */

	OBJECT(child)->state = OBJECT_ZOMBIE;
	object_destroy(wid);
}

